### ISSUE TITLE
Vulkan: FPS counter is not working on macOS due to MoltenVK missing hardware GPU timestamp support

### ISSUE THREAD

**Issue by RevoluPowered** [godotengine/godot#51371]
### Godot version

4.0.master

### System information

MacOS

### Issue description

Using the latest MoltenVK I'm not getting an FPS counter which works in master/4.0

I haven't ever seen it working.

### Steps to reproduce

1. Open master on Mac
2. Open viewport info
3. FPS counter will show the wrong values.

**Comment by RevoluPowered** [comment]
<img width="331" alt="Screenshot 2021-08-07 at 18 27 34" src="https://user-images.githubusercontent.com/748770/128608871-912e7c7c-41cb-435a-a889-f58c51f29c62.png">

**Comment by RevoluPowered** [comment]
![image](https://user-images.githubusercontent.com/748770/128608930-2264507e-ff8c-48d2-973a-ecf315cd6e6f.png)

**Comment by bruvzg** [comment]
Proper hardware timestamp support was added to the MoltenVK only 12 days ago - https://github.com/KhronosGroup/MoltenVK/pull/1404. Probably, previous software implementation of timestamps (which returns tick count) had insufficient precision to measure frame time.

**Comment by bruvzg** [comment]
With `master` (f9696449eae1b99dd7293db6285c21d35e51d539) version on MoltenVK it's showing something, but values seems to be wrong (empty scene is definitely running with more than 6 FPS):
<img width="653" alt="Screenshot 2021-08-07 at 21 24 49" src="https://user-images.githubusercontent.com/7645683/128610462-ad19b8c1-7610-4331-a80b-ff571130c8c7.png">

**Comment by Calinou** [comment]
> With `master` (f9696449eae1b99dd7293db6285c21d35e51d539) version on MoltenVK it's showing something, but values seems to be wrong (empty scene is definitely running with more than 6 FPS):

If the scene is effectively running at 650 FPS or so, then this might mean the value is multiplied by 1000.

**Comment by bruvzg** [comment]
> then this could mean the GPU time is erroneously being multiplied by 1000 on the MoltenVK side

It's probably not that simple. According to XCode Instruments profiler, average frame time is constant 16.68 ms (60 FPS, v-synced).

<img width="1129" alt="Screenshot 2021-08-07 at 21 39 58" src="https://user-images.githubusercontent.com/7645683/128610846-c6078d5f-7467-472d-879c-1f11787a6de0.png">

**Comment by RevoluPowered** [comment]
Nice so we can use the xcode instruments to measure frame time :) since its technically using Metal :D

**Comment by Calinou** [comment]
> It's probably not that simple. According to XCode Instruments profiler, average frame time is constant 16.68 ms (60 FPS, v-synced).

V-Sync is enabled by default (and perhaps forced by macOS too), but remember that this panel displays the time taken to render the frame on the GPU, not the time between two frames. This means that even with V-Sync enabled or FPS capped to a low value via low processor mode, it can report framerates *above* your monitor refresh rate

**Comment by bruvzg** [comment]
> Nice so we can use the xcode instruments to measure frame time :) since its technically using Metal :D

It should be usable for performance monitoring, and it's MoltenVK is transferring Vulkan pipeline function names and debug comment to the Metal side. It's not a RenderDoc, but should give an overall view.
<img width="882" alt="Screenshot 2021-08-07 at 21 56 55" src="https://user-images.githubusercontent.com/7645683/128611188-2d1aba70-84ac-4294-bf10-d194e06eea65.png">

**Comment by KoBeWi** [comment]
Is this issue still valid in the newest alpha?

**Comment by snougo** [comment]
> Is this issue still valid in the newest alpha?

Yes, it still valid in beta3, at least on my Intel GPU macbook pro

**Comment by snougo** [comment]
it still valid in beta5

**Comment by lostminds** [comment]
I just tried this on my mac (intel 4K iMac with Radeon Pro 560, macOS 13.0.1) and it seems to work fine now for me on 4.0.beta8

![Screenshot 2022-12-16 at 13 15 16](https://user-images.githubusercontent.com/17763524/208096378-6fad4ca3-bcc6-4fc0-847f-c558514c8752.jpg)

**Comment by akien-mga** [comment]
We've upgraded MoltenVK recently, including several months' work of improvements, so it might have fixed this.

Closing for now, please comment if you can still reproduce this issue in 4.0 beta 8 or later.

**Comment by snougo** [comment]
it still reproduced in beta9 on IntelGPU Mac

**Comment by lostminds** [comment]
It still works in beta9 on my mac intel-processor mac with a radeon GPU so it's not a general intel macOS issue (and maybe never was then). So the bug title could be misleading since "Intel macOS" is probably understood as a mac with an intel-based processor (as opposed to the newer M1 processors).

**Comment by Calinou** [comment]
Regarding Apple GPUs, see https://github.com/godotengine/godot/issues/96102#issuecomment-2311069062.
